src/snowflake/snowpark/mock/plan.py

```diff
@@ -808,11 +818,275 @@ def outer_join(base_df):
+                    )
+                    intermediate = join_result[condition]
+                else:
+                    intermediate = join_result
+
+                matched = target.apply(tuple, 1).isin(
+                    intermediate[target.columns].apply(tuple, 1)
+                )
+                matched.sf_type = ColumnType(BooleanType(), True)
+
+                # Update deleted row id set
+                for _, row in target[matched].iterrows():
+                    deleted_row_idx.add(row[ROW_ID])
+
+                # Delete rows in place
+                target = target[~matched]
+
+            elif isinstance(clause, InsertMergeExpression):
+                # calculate unmatched rows in the source
+                matched = source.apply(tuple, 1).isin(
+                    join_result[source.columns].apply(tuple, 1)
+                )
+                matched.sf_type = ColumnType(BooleanType(), True)
+                not_matched_rows = source[~matched]
+
+                # select unmatched rows that qualify the condition
+                if clause.condition:
+                    condition = calculate_expression(
+                        clause.condition, not_matched_rows, analyzer, expr_to_alias
+                    )
+                    not_matched_rows = not_matched_rows[condition]
+
+                # filter out the unmatched rows that have been inserted in previous clauses
+                not_matched_rows = not_matched_rows[
+                    ~not_matched_rows[SOURCE_ROW_ID].isin(inserted_row_idx).values
+                ]
+
+                # update inserted row idx set
+                for _, row in not_matched_rows.iterrows():
+                    inserted_row_idx.add(row[SOURCE_ROW_ID])
+
+                # Calculate rows to insert
+                rows_to_insert = TableEmulator(
+                    [], columns=target.drop(ROW_ID, axis=1).columns
+                )
+                rows_to_insert.sf_types = target.sf_types
+                if clause.keys:
+                    # Keep track of specified columns
+                    inserted_columns = set()
+                    for k, v in zip(clause.keys, clause.values):
+                        column_name = analyzer.analyze(k, expr_to_alias)
+                        inserted_columns.add(column_name)
+                        new_val = calculate_expression(
+                            v, not_matched_rows, analyzer, expr_to_alias
+                        )
+                        rows_to_insert[column_name] = new_val
+
+                    # For unspecified columns, use None as default value
+                    for unspecified_col in set(rows_to_insert.columns).difference(
+                        inserted_columns
+                    ):
+                        rows_to_insert[unspecified_col].replace(
+                            np.nan, None, inplace=True
+                        )
```

Comment:
knowledge question: what if users pass column name not in target table? shall we raise error?